### PR TITLE
Add : index date search params

### LINKED ISSUE #285: Handle date search parameters when the data type isn't date
the data type could be 

- date
- dateTime
- instant
- Period
- Timing

See https://www.hl7.org/fhir/search.html#date. At the moment we only index the date values. We need to handle the other 4 types of values.

### PR DESCRIPTION
**Pull request by epicadk**
Fixes #285

### CODE REVIEW

File: core/src/test/java/com/google/android/fhir/index/ResourceIndexerTest.kt [hunk 1]
```diff
@@ -431,5 +506,56 @@ class ResourceIndexerTest {
...
     const val FHIR_CURRENCY_SYSTEM = "urn:iso:std:iso:4217"
+
+    val TEST_OBSERVATION_EFFECTIVE_PERIOD = Observation()
+
+    init {
+      TEST_OBSERVATION_EFFECTIVE_PERIOD.apply {
+        id = "non-null ID"
+        effective =
+          Period().apply {
+            start = Date(1516055094)
```
Comment (deepankarb) [inline core/src/test/java/com/google/android/fhir/index/ResourceIndexerTest.kt:517]: Can we add the values in any human readable form? E.g.:
```suggestion
            start = Date(1516055094) // Tue Jan 16 03:54:54 IST 2018
```

**Review by deepankarb — CHANGES_REQUESTED** [review]
Thanks for adding this :) I have left some comments, please fix and I'll approve.

File: core/src/main/java/com/google/android/fhir/index/ResourceIndexer.kt [hunk 2]
```diff
@@ -132,6 +135,40 @@ internal object ResourceIndexer {
...
       }
+      "Period" -> {
+        val period = value as Period
+        DateIndex(
+          searchParam.name,
+          searchParam.path,
+          if (period.hasEnd()) period.end.time else Long.MAX_VALUE,
+          if (period.hasStart()) period.start.time else Long.MIN_VALUE,
+          when {
+            (period.hasEnd()) -> period.endElement.precision
```
[the remark this answers is not in the document]
Comment (epicadk) [reply core/src/main/java/com/google/android/fhir/index/ResourceIndexer.kt:148]: I tried this. It should work but I know it probably isn't the best way to do this. Do you have any suggestions?

**Comment by epicadk** [comment]
@deepankarb made the request changes can you review it again?

**Comment by epicadk** [comment]
@deepankarb gentle reminder can you please review the pr again? 😅

**Review by deepankarb — APPROVED** [review]
Thanks!

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/google/android/fhir/index/ResourceIndexer.kt [hunk 3]
```diff
@@ -135,6 +138,42 @@ internal object ResourceIndexer {
...
+          when {
+            (period.hasEnd() and period.hasStart()) ->
+              maxOf(period.startElement.precision, period.endElement.precision)
+            (period.hasEnd()) -> period.endElement.precision
+            (period.hasStart()) -> period.startElement.precision
+            else -> DateTimeType.DEFAULT_PRECISION
+          }
+        )
+      }
+      "Timing", "Timing.repeat" -> {
```
[the remark this answers is not in the document]
Comment (epicadk) [reply core/src/main/java/com/google/android/fhir/index/ResourceIndexer.kt:157]: Sorry you're right Timing.repeat is the fhirtype for the Timing repeat component. Removing.

File: core/src/main/java/com/google/android/fhir/index/ResourceIndexer.kt [hunk 4]
```diff
@@ -135,6 +138,42 @@ internal object ResourceIndexer {
           instant.precision
         )
       }
+      "Period" -> {
```
[the remark this answers is not in the document]
Comment (epicadk) [reply core/src/main/java/com/google/android/fhir/index/ResourceIndexer.kt:151]: Made the changes 👍🏼

**Review by jingtang10 — APPROVED** [review]
Thanks again for your contribution! 👍

**Pull request merged**